src/common/mfu_daos.c

```diff
@@ -610,6 +610,300 @@ static int daos_any_error(int rank, bool local_daos_error, int flag_daos_args)
+        rc = 1;
+        goto out;
+    }
+
+    /* Fetch the ACL separately in case user doesn't have access */
+    rc = daos_cont_get_acl(coh, &prop_acl, NULL);
+    if (rc == 0) {
+        /* ACL will be appended to the end */
+        props_merged = daos_prop_merge(props, prop_acl);
+        if (props_merged == NULL) {
+            MFU_LOG(MFU_LOG_ERR, "Failed to set container ACL: "DF_RC, DP_RC(rc));
+            rc = 1;
+            goto out;
+        }
+        daos_prop_free(props);
+        props = props_merged;
+    } else if (rc && rc != -DER_NO_PERM) {
+        MFU_LOG(MFU_LOG_ERR, "Failed to query container ACL: "DF_RC, DP_RC(rc));
+        rc = 1;
+        goto out;
+    }
+
+    rc = 0;
+    *_props = props;
+
+out:
+    daos_prop_free(prop_acl);
+    if (rc != 0) {
+        daos_prop_free(props);
+    }
+
+    return rc;
+}
+
+/*
+ * Free the user attribute buffers created by cont_get_usr_attrs.
+ */
+static void cont_free_usr_attrs(int n, char*** _names, void*** _buffers,
+                               size_t** _sizes)
+{
+    char**  names = *_names;
+    void**  buffers = *_buffers;
+
+    if (names != NULL) {
+        for (size_t i = 0; i < n; i++) {
+            mfu_free(&names[i]);
+        }
+        mfu_free(_names);
+    }
+    if (buffers != NULL) {
+        for (size_t i = 0; i < n; i++) {
+            mfu_free(&buffers[i]);
+        }
+        mfu_free(_buffers);
+    }
+    mfu_free(_sizes);
+}
+
+/*
+ * Get the user attributes for a container in a format similar
+ * to what daos_cont_set_attr expects.
+ * cont_free_usr_attrs should be called to free the allocations.
+ * Returns 1 on error, 0 on success.
+ */
+static int cont_get_usr_attrs(daos_handle_t coh,
+                              int* _n, char*** _names, void*** _buffers,
+                              size_t** _sizes)
+{
+    int         rc = 0;
+    uint64_t    total_size = 0;
+    uint64_t    cur_size = 0;
+    uint64_t    num_attrs = 0;
+    uint64_t    name_len = 0;
+    char*       name_buf = NULL;
+    char**      names = NULL;
+    void**      buffers = NULL;
+    size_t*     sizes = NULL;
+
+    /* Get the total size needed to store all names */
+    rc = daos_cont_list_attr(coh, NULL, &total_size, NULL);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to list user attributes "DF_RC, DP_RC(rc));
+        rc = 1;
+        goto out;
+    }
+
+    /* no attributes found */
+    if (total_size == 0) {
+        *_n = 0;
+        goto out;
+    }
+
+    /* Allocate a buffer to hold all attribute names */
+    name_buf = MFU_CALLOC(total_size, sizeof(char));
+    if (name_buf == NULL) {
+        MFU_LOG(MFU_LOG_ERR, "failed to allocate user attribute buffer");
+        rc = 1;
+        goto out;
+    }
+
+    /* Get the attribute names */
+    rc = daos_cont_list_attr(coh, name_buf, &total_size, NULL);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to list user attributes "DF_RC, DP_RC(rc));
+        rc = 1;
+        goto out;
+    }
```

Comment:
Just curious about the ``daos_cost_list_attr`` interface.  Is ``total_size`` used as an input parameter here to specify the buffer size?

If so, I'm guessing this function might return an error if the attribute list size has increased, since the first time this was called.

And I suspect it's valid for the attribute list size to be less than what it was?

Reply:
@adammoody I believe it is valid for it to be less than it was, but I will note that we should return an error if the attribute size has increased since the first time it was called.

Reply:
@adammoody @dsikich Yes, it is possible for the actual size to change between the first call to `daos_cont_list_attr` (which just gets the size) and the second call (which gets the names).
https://github.com/daos-stack/daos/blob/011b78cb28aaf162af635f37d1fcabc62076ef0b/src/include/daos_cont.h#L382-L399
I **_think_** this is how it works, based on other DAOS paradigms:
- If size is less than the size passed in
  - The output size set accordingly, since it's passed by reference, and the buffer contains all the data
- If size is more than the size passed in
  - The output size is set accordingly, and the buffer is **_truncated_**

The "foolproof" way to handle this would be to have a retry loop, and realloc and fetch every time the size is **_more_** than the size of the buffer.